template-ssr-react-streaming-ts/server.js

```diff
@@ -0,0 +1,100 @@
+import fs from 'node:fs/promises'
+import express from 'express'
+
+// Constants
+const isProduction = process.env.NODE_ENV === 'production'
+const port = process.env.PORT || 5173
+const base = process.env.BASE || '/'
+const ABORT_DELAY = 10000
+
+// Cached production assets
+const templateHtml = isProduction
+  ? await fs.readFile('./dist/client/index.html', 'utf-8')
+  : ''
+const ssrManifest = isProduction
+  ? await fs.readFile('./dist/client/.vite/ssr-manifest.json', 'utf-8')
+  : undefined
+
+// Create http server
+const app = express()
+
+// Add Vite or respective production middlewares
+let vite
+if (!isProduction) {
+  const { createServer } = await import('vite')
+  vite = await createServer({
+    server: { middlewareMode: true },
+    appType: 'custom',
+    base
+  })
+  app.use(vite.middlewares)
+} else {
+  const compression = (await import('compression')).default
+  const sirv = (await import('sirv')).default
+  app.use(compression())
+  app.use(base, sirv('./dist/client', { extensions: [] }))
+}
+
+// Serve HTML
+app.use('*', async (req, res) => {
+  try {
+    const url = req.originalUrl.replace(base, '')
+
+    let template
+    let render
+    if (!isProduction) {
+      // Always read fresh template in development
+      template = await fs.readFile('./index.html', 'utf-8')
+      template = await vite.transformIndexHtml(url, template)
+      render = (await vite.ssrLoadModule('/src/entry-server.tsx')).render
+    } else {
+      template = templateHtml
+      render = (await import('./dist/server/entry-server.js')).render
+    }
+
+    let didError = false
+
+    const { pipe, abort } = render(url, ssrManifest, {
+      onShellError() {
+        res.status(500)
+        res.set({ 'Content-Type': 'text/html' })
+        res.send('<h1>Something went wrong</h1>')
+      },
+      onAllReady() {
```

Comment:
If I understand correctly, `onAllReady` will cause React to wait for everything to finish render before responding. Wouldn't it be simpler in that case to not stream? Part of the benefit of streaming is that HTML can be slowly loaded to the client, but this will buffer the entire HTML before sending to the client.

And if we do true streaming, `helmet.title.toString()` wouldn't work as we're not rendering all the components beforehand. I think we can omit that information for now and is up to the consumer to structure the head streaming itself.

Reply:
We can use `onShellReady` here, ok.